lib/api/ldp/api.js

```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+const { LdpOperation } = require('./ldp-operation')
+const HttpError = require('standard-http-error')
+
+class LdpHttpHandler {
+  /**
+   * Usage:
+   *   ```
+   *   const store = new LdpFileStore({host, fetch, ...})
+   *   const acl = new AclChecker({host, fetch, store, ...})
+   *
+   *   const ldp = new LdpHttpHandler({store, acl})
+   *   ```
+   *
+   * @param options.store {LdpFileStore|LdpMemoryStore|LdpQuadStore} storage backend
+   *   (either file system based, or in-memory, or SPARQL based quad store, etc)
+   *
+   * @param options.acl {AclChecker} ACL checker component
+   *
+   * @param options.host {SolidHost} Server config object
+   */
+  constructor (options) {
+    this.store = options.store
+    this.acl = options.acl
+    this.host = options.host
+  }
+
+  /**
+   * Provides an Express handler for LDP API requests.
+   * Authentication has already happened earlier in the middleware stack, stored
+   * in session.
+   * Usage:
+   *   ```
+   *   app.use('/', ldp.handleRequest)
+   *   ```
+   *
+   * @see https://github.com/solid/solid-architecture/blob/master/server/request-flow.md
+   *
+   * @throws {Error} 401 If request not authenticated but resource is non-public
+   *
+   * @throws {Error} 404 If resource is not found, OR if found and request
+   *   is not authorized to access it (default behavior, can be overridden by
+   *   owner of resource)
+   *
+   * @throws {Error} 403 If request is not authorized to access resource and
+   *   user has enabled "Request permission" action
+   *
+   * @throws {Error} 400 If invalid parameters (or error parsing request body,
+   *   for cases like PATCH requests)
+   *
+   * @throws {Error} 409 If PATCH operation results in conflict
+   *
+   * @throws {Error} 406 If no appropriate representation found (for content type)
+   *
+   * @throws {Error} 405 If HTTP method not allowed / not implemented
+   */
+  async handleRequest (req, res, next) {
+    try {
+      const operation = this.operationFrom({req, store: this.store})
+
+      // check that operation is permitted (throws error if not)
+      const permissions = await this.acl.allow(operation)
+
+      // perform the operation and return a response
+      const result = await operation.perform({res, store: this.store})
+
+      // write both generic and op-specific headers
+      // note: you need `permissions` for the WAC-Allow header
+      operation.writeHeaders({res, result, permissions})
+
+      operation.writeResponse({res, result})
+    } catch (error) {
+      next(error)
+    }
+  }
+
+  async operationFrom (req) {
+    const options = this.parseOperation(req)
+
+    const Operation = LdpOperation.BY_METHOD[options.method]
```

Comment:
@RubenVerborgh ok, so this is the next iteration based on our discussion.

The factory code was move from `Operation.from()` static method, to this instance factory method on `ldp`.

Here's my main problem with this design, though (specifically, getting the operation by HTTP method).

It breaks down when it comes to the GET method. (And PATCH method to a certain extent.) But let's talk about the get.

For most methods, there's a 1:1 mapping of HTTP method to Operation. But for GET, it's more like 1 get method to 3-4 potential operations. Where does the logic dispatching between them belong?

Reply:
Somewhere there needs to be a `switch()` statement :)

Reply:
> But for GET, it's more like 1 get method to 3-4 potential operations.

What are they?

Reply:
So really, this line should be more like:

```js
if (method === 'get') {
  Operation = this.parseGetOperation(...)
}  else {
  Operation = LdpOperation.BY_METHOD[method]
}
```

Reply:
Re : cases for GET (off the top of my head)

- Resource
- Container
- Globbing
- Streaming media ie byte ranges
- Data browser
- Conneg

Reply:
Wow, that's a lot more than I thought. Conneg is orthonogal though, but the rest of them seem individual cases indeed.

So to get back to the point then, I'd suggest `Operation = new LdpOperation.BY_METHOD[method]({ method, target, body, … })` or similar, such that `LdpGetOperation` is responsible for creating its own subtypes.

Reply:
>  such that LdpGetOperation is responsible for creating its own subtypes.

Ah, wait tho - that would mean instantiating different subtypes in the constructor of LdpGetOperation.. Which is awkward, no? Maybe a static `from()` would work? :)

Reply:
I suppose, part of this depends on whether we want to model "List container" / "Glob operation" / etc as separate LdpOperation subclasses...

Reply:
> Ah, wait tho - that would mean instantiating different subtypes in the constructor of LdpGetOperation.. Which is awkward, no?

Not necessarily. Can be factory too, with our without `from` :smile: The difference is purely syntactical.

> I suppose, part of this depends on whether we want to model "List container" / "Glob operation" / etc as separate LdpOperation subclasses...

Indeed. It's a manner of how we want to manage complexity. Could be through subclasses, could be through different methods in LdpGetOperation. It's not that big of a decision, I think. Important thing is that the API shouldn't know how to construct different kinds of operations, that knowledge should be with the operations themselves.

Reply:
Sounds good :)